vignettes/evals.Rmd

```diff
@@ -0,0 +1,180 @@
+date: "`r Sys.Date()`"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{Evals}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+```{r, echo = FALSE, message = FALSE}
+knitr::opts_chunk$set(collapse = T, comment = "#>")
+library(pander)
+```
+
+`evals` is aimed at collecting as much information as possible while evaluating R code. It can evaluate a character vector of R expressions, and it returns a list of captured information while running those:
+
+* `src` holds the R expression,
+* `result` contains the raw R object as is,
+* `output` represents how the R object is printed to the standard output,
+* `type` is the class of the returned R object,
+* `msg` is a list of possible messages captured while running the R expression and
+* `stdout` contains if anything was written to the standard output.
+
+Besides capturing evaluation information, `evals` is able to automatically identify if an R expression is returning anything to a graphical device, and can save the resulting image in a variety of file formats.
+
+Another interesting feature that `evals` provides is caching the results of evaluated expressions. For more details read the section about Caching.
+
+## Basic Usage
+
+Let's start with basic example of evaluating `1:10` and collecting all information about it:
+
+```{r}
+evals('1:10')
+```
+
+It is possible to capture only some of the information, by specifying `output` parameter:
+
+```{r}
+evals('1:10', output = c('result', 'output'))
+```
+
+### Errors/Warings
+```{r}
+evals('x')[[1]]$msg
+```
+
+### Graphs and Graphical Options
+
+### Style unification
+
+## Hooks
+
+`evals` provdes facilities to call user specified functions (hooks) for specific types of objects. It can be achieved by specifying `hooks` parameter with named list where names are object types and elements are functions to be called. For example image, you want to automatically `round` all `numeric` objects:
+
+```{r}
+evals('1:10/10', hooks = list('numeric' = round))[[1]]$result
+```
+
+Or `unlist` a `list`:
+
+```{r}
+evals('as.list(1:10)', hooks = list('list' = unlist))[[1]]$result
+```
+
+Hooks in `evals` are implemented using `do.call` logic, which also allows to supply parameters to each hook by making elements of `hooks` parameter lists where arguments start at second position. Extending previous examples:
+
+```{r}
+evals('1:10/10', hooks = list('numeric' = list(round, 1)))[[1]]$result
+```
+
+Or
+
+```{r}
+hooks <- list('numeric' = list(round, 2), 'matrix' = list(round, digits=1))
+res <- evals('1:10/1000; matrix(1:10/100, nrow=2)', hooks = hooks)
+res[[1]]$result
+res[[2]]$result
+```
+
+## Logging
+
+To make execution and debug more understandable `evals` provides a possibility for logging with `log` parameter. Logging in `evals` relies on [futile.logger](cran.r-project.org/web/packages/futile.logger/futile.logger.pdf) package which provides logging API similar to `log4j`. Basic example:
```

Comment:
This is especially useful if you are using `Pandoc.brew`, see e.g. the demo in http://blog.rapporter.net/2014/10/pander-050-next-generation-of-markdown.html